### PR TITLE
Add greenlet recipe.

### PR DESCRIPTION
**Pull request by curtisalexander**
Fixes https://github.com/conda-forge/staged-recipes/issues/2943

### CODE REVIEW

**Comment by curtisalexander** [comment]
Unsure how to proceed at this point.  Throwing the following error when the build starts...

```
BUILD START (revised): greenlet-0.4.12-py27_0
Package: greenlet-0.4.12-py27_0
source tree in: /Users/travis/miniconda/conda-bld/greenlet_1492302817950/work/greenlet-0.4.12
+ source /Users/travis/miniconda/bin/activate /Users/travis/miniconda/conda-bld/greenlet_1492302817950/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p
+ python setup.py install --single-version-externally-managed --record record.txt
usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
   or: setup.py --help [cmd1 cmd2 ...]
   or: setup.py --help-commands
   or: setup.py cmd --help
error: option --single-version-externally-managed not recognized
Traceback (most recent call last):
  File "/Users/travis/miniconda/bin/conda-build-all", line 6, in <module>
    sys.exit(conda_build_all.cli.main())
  File "/Users/travis/miniconda/lib/python3.6/site-packages/conda_build_all/cli.py", line 90, in main
    b.main()
  File "/Users/travis/miniconda/lib/python3.6/site-packages/conda_build_all/builder.py", line 267, in main
    built_dist_location = self.build(meta, build_config)
  File "/Users/travis/miniconda/lib/python3.6/site-packages/conda_build_all/builder.py", line 204, in build
    conda_build.api.build(meta.meta, config=config)
  File "/Users/travis/miniconda/lib/python3.6/site-packages/conda_build/api.py", line 97, in build
    need_source_download=need_source_download, config=config)
  File "/Users/travis/miniconda/lib/python3.6/site-packages/conda_build/build.py", line 1503, in build_tree
    config=config)
  File "/Users/travis/miniconda/lib/python3.6/site-packages/conda_build/build.py", line 1135, in build
    utils.check_call_env(cmd, env=env, cwd=src_dir)
  File "/Users/travis/miniconda/lib/python3.6/site-packages/conda_build/utils.py", line 620, in check_call_env
    return _func_defaulting_env_to_os_environ(subprocess.check_call, *popenargs, **kwargs)
  File "/Users/travis/miniconda/lib/python3.6/site-packages/conda_build/utils.py", line 616, in _func_defaulting_env_to_os_environ
    return func(_args, **kwargs)
  File "/Users/travis/miniconda/lib/python3.6/subprocess.py", line 291, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/bin/bash', '-x', '-e', '/Users/travis/miniconda/conda-bld/greenlet_1492302817950/work/greenlet-0.4.12/conda_build.sh']' returned non-zero exit status 1.
The command "if [ -n "$GH_TOKEN" ]; then echo "Creating feedstocks from the recipe(s)."; git config --global user.name "Travis-CI on github.com/conda-forge/staged-recipes"; git config --global user.email "conda-forge@googlegroups.com"; source ./.CI/create_feedstocks; else echo "Building all recipes."; source ./.CI/build_all; fi" exited with 1.
Done. Your build exited with 1.
```

**Comment by chrisburr** [comment]
I had the same issue with #2833. It seems setuptools has to be imported in `setup.py` to monkey patch distutils to having this option, I did this using sed on [line 5 here](https://github.com/conda-forge/staged-recipes/pull/2833/files#diff-bf27819aed6c5d0c433b3a39d9a92e38) but there might be a better solution.

**Comment by curtisalexander** [comment]
@chrisburr I researched how to use `sed` to change `setup.py`.  Unfortunately, I want to be able to build for Windows as well and would need to come up with something different to make the needed adjustment.

**Comment by chrisburr** [comment]
@curtisalexander I haven't tried it but you should be able to apply the fix to the source using a by making a patch file[1]. I assume conda will then apply it in a platform independent way.

[1] https://conda.io/docs/building/meta-yaml.html#patches

**Comment by curtisalexander** [comment]
@chrisburr Thank you for the suggestion.  I will have to research an example of patch file and test out.

**Comment by curtisalexander** [comment]
If I cannot get the patch file to work per Chris' suggestion, then I probably could use some guidance on how to correct without overwriting the `setup.py` file.

The issue seems to be within `setup.py`.  There is a check for `setuptools` and if the import fails, then the function `setup` from `distutils` is used instead of `setup` from `setuptools`.

Below is the relevant snippet from the greenlet [`setup.py`](https://github.com/python-greenlet/greenlet/blob/master/setup.py#L16-L28) file.

```{python}
try:
    if not (sys.modules.get("setuptools")
            or "develop" in sys.argv
            or "upload" in sys.argv
            or "bdist_egg" in sys.argv
            or "bdist_wheel" in sys.argv
            or "test" in sys.argv):
        raise ImportError()
    from setuptools import setup, Extension
    setuptools_args = dict(test_suite='tests.test_collector', zip_safe=False)
except ImportError:
    from distutils.core import setup, Extension
    setuptools_args = dict()
```

In fact, I have no real need or use for `greenlet` other than trying to add all the dependencies for [`python-neovim`](https://github.com/conda-forge/staged-recipes/pull/2836).

File: recipes/greenlet/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  version: {{ version }}
+
+source:
+  fn: {{ name }}-{{ version }}.tar.gz
+  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
+  sha256: {{ sha256 }}
+
+build:
+  number: 0
+  script: python setup.py install --single-version-externally-managed --record record.txt
```
Comment (jakirkham) [inline recipes/greenlet/meta.yaml]: Please drop `--single-version-externally-managed --record record.txt`. These are only needed if one uses `setuptools`, which it appears is not the case here.

**Comment by jakirkham** [comment]
I think once the noted change is made this recipe will build fine.

File: recipes/greenlet/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    - greenlet
+
+about:
+  home: https://github.com/python-greenlet/greenlet
+  license: MIT
+  license_family: MIT
+  license_file: LICENSE
+  summary: 'Lightweight in-process concurrent programming'
+
+  description: |
```
Comment (stuertz) [inline recipes/greenlet/meta.yaml]: You should replace the ticks and doublequotes in the description with ASCII quotes and doublequotes. This fixes the code problem with py35/py36 appveyor builds

Comment (curtisalexander) [inline recipes/greenlet/meta.yaml]: @stuertz Thank you - just re-submitted.

File: recipes/greenlet/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  name: {{ name|lower }}
+  version: {{ version }}
+
+source:
+  fn: {{ name }}-{{ version }}.tar.gz
+  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
+  sha256: {{ sha256 }}
+
+build:
+  number: 1
```
Comment (jakirkham) [inline recipes/greenlet/meta.yaml]: Please reset to `0`. All recipes at staged-recipes should start with build number `0`.

Note: When changing the version (to an unreleased one), one should reset to the build number to `0` as well.

File: recipes/greenlet/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  home: https://github.com/python-greenlet/greenlet
+  license: MIT
+  license_family: MIT
+  license_file: LICENSE
+  summary: 'Lightweight in-process concurrent programming'
+
+  description: |
+    The greenlet package is a spin-off of Stackless, a version of CPython that supports micro-threads called "tasklets". Tasklets run pseudo-concurrently (typically in a single or a few OS-level threads) and are synchronized with data exchanges on "channels".
+
+    A "greenlet", on the other hand, is a still more primitive notion of micro-thread with no implicit scheduling; coroutines, in other words. This is useful when you want to control exactly when your code runs. You can build custom scheduled micro-threads on top of greenlet; however, it seems that greenlets are useful on their own as a way to make advanced control flow structures. For example, we can recreate generators; the difference with Python's own generators is that our generators can call nested functions and the nested functions can yield values too. Additionally, you don't need a "yield" keyword. See the example in tests/test_generator.py.
```
Comment (jakirkham) [inline recipes/greenlet/meta.yaml]: Might wrap this text a bit. It gets a bit long reading horizontally. We don't have a hard line limit, but feel free to use your best judgement. Let's say not having to scroll horizontally would be a good goal.

**Comment by jakirkham** [comment]
Made a couple notes, but should be pretty easy to address. Please ping when they are done.

File: recipes/greenlet/meta.yaml (hunk 4 above)
Comment (curtisalexander) [inline recipes/greenlet/meta.yaml]: Good point...how about the [magical](https://www.python.org/dev/peps/pep-0008/#maximum-line-length) 79 character limit.

File: recipes/greenlet/meta.yaml (hunk 3 above)
Comment (curtisalexander) [inline recipes/greenlet/meta.yaml]: Reverting.

File: recipes/greenlet/meta.yaml (hunk 4 above)
Comment (jakirkham) [inline recipes/greenlet/meta.yaml]: Sure, that seems fine. Thanks.

**Pull request merged**